content/blog/2023-03-17-build-grafana-dashboards-for-apache-skywalking-native-promql-support/index.md

```diff
@@ -0,0 +1,403 @@
+- Entity: represents the metrics' belonging and used for the query.
+  An Entity will contain the following information depending on the `Scope`：
+  Scope represents the metrics level and in query stage represents the Scope catalog,
+  Scope catalog provides high-dimension classifications for all scopes as a hierarchy structure.
+
+| Scope                   | Entity Info                            |
+|-------------------------|----------------------------------------|
+| Service                 | Service(include layer info)            |
+| ServiceInstance         | Service, ServiceInstance               |
+| Endpoint                | Service, Endpoint                      |
+| ServiceRelation         | Service, DestService                   |
+| ServiceInstanceRelation | ServiceInstance, DestServiceInstance   |
+| EndpointRelation        | Endpoint, DestEndpoint                 |
+| Process                 | Service, ServiceInstance, Process      |
+| ProcessRelation         | Process,  ServiceInstance, DestProcess |
+
+- Value:
+1. single value: long.
+2. labeled value: text, `label1,value1|label2,value2|...`, such as `L2 aggregation,5000 | L1 aggregation,8000`.
+
+- TimeBucket: the time is accurate to minute, hour, day.
+
+## How to use PromQL Service
+
+### Setup
+PromQL Service is enabled by default after v9.4.0, so no additional configuration is required.
+The default ports, for example, can be configured by using OAP environment variables:
+```text
+restHost: ${SW_PROMQL_REST_HOST:0.0.0.0}
+restPort: ${SW_PROMQL_REST_PORT:9090}
+restContextPath: ${SW_PROMQL_REST_CONTEXT_PATH:/}
+restMaxThreads: ${SW_PROMQL_REST_MAX_THREADS:200}
+restIdleTimeOut: ${SW_PROMQL_REST_IDLE_TIMEOUT:30000}
+restAcceptQueueSize: ${SW_PROMQL_REST_QUEUE_SIZE:0}
+```
+
+### Use Prometheus expression
+PromQL is matched metric through the Prometheus expression, a typical Prometheus metric is as follows:
+
+![Prometheus metric](img_2.png#pic_left)
+
+
+For the match, the Prometheus expression is as follows:
+
+![Prometheus expression](img_4.png#pic_left)
+
+
+So in the PromQL Service, the expression will be parsed to a metric name and entity info for query, the mapping is as follows:
+
+| SkyWalking Concepts | Prometheus expression   |
+|---------------------|-------------------------|
+| Metric name         | Metric name             |
+| Layer               | Label<layer>            |
+| Service             | Label<service>          |
+| ServiceInstance     | Label<service_instance> |
+| Endpoint            | Label<endpoint>         |
+| ...                 | ...                     |
+
+For example, the following expressions are used to match query metrics: service_cpm, service_instance_cpm, endpoint_cpm
+
+```text
+service_cpm{service='agent::songs', layer='GENERAL'}
+service_instance_cpm{service='agent::songs', service_instance='agent::songs_instance_1', layer='GENERAL'}
+endpoint_cpm{service='agent::songs', endpoint='GET:/songs', layer='GENERAL'} 
```

Comment:
Could you add comments for these expressions to explain the query? And parsed GraphQL for a better understanding.